#!/bin/bash
#SBATCH -J run_gwas_dosage
#SBATCH --mail-type=BEGIN,END,FAIL
#SBATCH --mail-user=c.ke@lumc.nl
#SBATCH -N6
#SBATCH --ntasks=6
#SBATCH --mem-per-cpu=8096MB
#SBATCH --time=48:00:00


###########
#  GWAS with dosage, run on plink2 files generated by Mingdong
#  Script run with the input data settings:
#  - Imputated data used for all cohorts (was fully complete)
#  - NO LD PRUNING 
#  - Remission yes/no of 6 months
#  - made by: Samantha Jurado Zapata
#  - current version modified by: Mingdong Liu, Changlin Ke
###########

cd /exports/reum/mliu/GWAS_dosage/all_cohorts/

#/exports/reum/knevel_lab/plink2.3/plink2 --pfile EARTH/SetA/EA127genotypes_merged --glm hide-covar --ci 0.95 --covar /exports/reum/knevel_lab/Scripts_Samantha/GWAS_run/Version8/EARTH_A/earth_A.covar --covar-variance-standardize --adjust --out EARTH/SetA/gwas_dosage_EARTHA

/exports/reum/knevel_lab/plink2.3/plink2 --pfile EARTH/SetB/EARTH_setB_merged --glm hide-covar --ci 0.95 --covar EARTH/SetB/EARTH_setB_merged.covar --covar-variance-standardize --adjust --out EARTH/SetB/gwas_dosage_EARTHB

/exports/reum/knevel_lab/plink2.3/plink2 --pfile RAMS/RAMS_merged --glm hide-covar --ci 0.95 --covar RAMS/RAMS.covar --covar-variance-standardize --adjust --out RAMS/gwas_dosage_RAMS

/exports/reum/knevel_lab/plink2.3/plink2 --pfile MADRID/madrid_merged --glm hide-covar --ci 0.95 --covar MADRID/madrid.covar --covar-variance-standardize --adjust --out MADRID/gwas_dosage_madird

/exports/reum/knevel_lab/plink2.3/plink2 --pfile EIRA/EIRA_merged --glm hide-covar --ci 0.95 --covar EIRA/EIRA.covar --covar-variance-standardize --adjust --out EIRA/gwas_dosage_EIRA

/exports/reum/knevel_lab/plink2.3/plink2 --pfile PEAC/PEAC_merged --glm hide-covar --ci 0.95 --covar PEAC/PEAC.covar --covar-variance-standardize --adjust --out PEAC/gwas_dosage_PEAC

/exports/reum/knevel_lab/plink2.3/plink2 --pfile ACPA-/ACPAneg_merged --glm hide-covar --ci 0.95 --covar ACPA-/ACPA.covar --covar-variance-standardize --adjust --out ACPA-/gwas_dosage_ACPA

/exports/reum/knevel_lab/plink2.3/plink2 --pfile SERA/sera_merged --glm hide-covar --ci 0.95 --covar SERA/sera.covar --covar-variance-standardize --adjust --out SERA/gwas_dosage_SERA

/exports/reum/knevel_lab/plink2.3/plink2 --pfile EAC/EAC_merged --glm hide-covar --ci 0.95 --covar EAC/EAC.covar --covar-variance-standardize --adjust --out EAC/gwas_dosage_EAC

/exports/reum/knevel_lab/plink2.3/plink2 --pfile VIENNA/vienna_merged --glm hide-covar --ci 0.95 --covar VIENNA/vienna.covar --covar-variance-standardize --adjust --out VIENNA/gwas_dosage_vienna

#run with vcf files
#/exports/reum/knevel_lab/plink2.3/plink2 --vcf all_cohorts_merged_v2.vcf.gz dosage=HDS --glm hide-covar --ci 0.95 --covar all_covar_v2.covar --covar-variance-standardize --pheno cohorts_merged_v2.pheno --out gwas_v2




